src/sdk/namespace_fs.js

```diff
@@ -790,12 +806,28 @@ class NamespaceFS {
             const log2_size_histogram = {};
             let drain_promise = null;
 
-            dbg.log0('NamespaceFS: read_object_stream', { file_path, start, end });
+            dbg.log0('NamespaceFS: read_object_stream', {
+                file_path, start, end, size: stat.size,
+            });
 
             let count = 1;
             for (let pos = start; pos < end;) {
                 object_sdk.throw_if_aborted();
 
+                // Our buffer pool keeps large buffers and we want to avoid spending
+                // all our large buffers and then have them waiting for high latency calls
+                // such as reading from archive/on-demand cache files.
+                // Instead, we detect the case where a file is "sparse",
+                // and then use just a small buffer to wait for a tiny read,
+                // which will recall the file from archive or load from remote into cache,
+                // and once it returns we can continue to the full fledged read.
+                if (is_sparse_file(stat)) {
+                    dbg.log0('NamespaceFS: read_object_stream - warmup sparse file', {
+                        file_path, pos, size: stat.size, blocks: stat.blocks,
+                    });
+                    await file.read(fs_context, this.warmup_buffer, 0, 1, pos);
+                }
+
```

Comment:
@eshelmarc This is the logic to improve buffer pool usage - 
I only read 1 byte on the current pos, so it assumes that prefetching will get a whole 8MB.